bitbucket-slack-server-integration-plugin/src/main/java/com/atlassian/bitbucket/plugins/slack/listener/BitbucketNotificationEventListener.java

```diff
@@ -101,11 +102,26 @@ public void onEvent(final PullRequestEvent event) {
                             getUsersAddedToPullRequest(event),
                             false),
                     options -> {
+                        final boolean isReviewersUpdate = (event instanceof PullRequestOpenedEvent && !event.getPullRequest().getReviewers().isEmpty()) ||
+                                event instanceof PullRequestReviewersUpdatedEvent;
                         if (isReviewersUpdate) {
+                            Set<ApplicationUser> addedReviewers;
+                            Set<ApplicationUser> removedReviewers;
+                            if (event instanceof PullRequestOpenedEvent) {
+                                addedReviewers = participantsToApplicationUser(event.getPullRequest().getReviewers());
+                                removedReviewers = Collections.emptySet();
+                            } else {
+                                PullRequestReviewersUpdatedEvent reviewersUpdatedEvent = (PullRequestReviewersUpdatedEvent) event;
+                                addedReviewers = reviewersUpdatedEvent.getAddedReviewers();
+                                removedReviewers = reviewersUpdatedEvent.getRemovedReviewers();
+                            }
+                            final boolean hasUserAddedOrRemovedHimself = hasUserAddedOrRemovedHimself(event.getUser(),
+                                    addedReviewers, removedReviewers);
                             if (hasUserAddedOrRemovedHimself || options.isPersonal()) {
-                                return ofNullable(slackNotificationRenderer.getReviewersPullRequestMessage(
-                                        (PullRequestReviewersUpdatedEvent) event,
-                                        Verbosity.EXTENDED.equals(options.getVerbosity())));
+                                for (ApplicationUser user : addedReviewers) {
+                                    return ofNullable(slackNotificationRenderer.getReviewersPullRequestMessage(
+                                            Verbosity.EXTENDED.equals(options.getVerbosity()), event.getPullRequest(), user, addedReviewers));
```

Comment:
`return` in a loop actually makes it useless. Only first iteration will be executed.
I feel like in this case multiple notifications may need to be sent, that isn't supported by current architecture, so it needs to be updated. Instead of `Function<NotificationRenderingOptions, Optional<ChatPostMessageRequestBuilder>>` this lambda should be of the following type: `Function<NotificationRenderingOptions, List<ChatPostMessageRequestBuilder>>`

Reply:
Good call, I haven't notice the `return` here.
So, this is a channel notification and the notification should not be sent once for the channel.
There's no need to iterate over the added reviewers because the array is already passed to formatting function a couple of lines below.